internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -499,6 +509,38 @@ func (r *BMACReconciler) reconcileNodeLabels(bmh *bmh_v1alpha1.BareMetalHost, ag
 	return false
 }
 
+func (r *BMACReconciler) reconcileClusterReference(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, agent *aiv1beta1.Agent, log logrus.FieldLogger) (bool, error) {
+	if clusterReferenceStr, ok := bmh.ObjectMeta.Annotations[BMH_CLUSTER_REFERENCE]; ok {
+		infraEnv, err := r.findInfraEnvForBMH(ctx, log, bmh)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to find infraenv for bmh %s/%s", bmh.Namespace, bmh.Name)
+		}
+		if infraEnv != nil && infraEnv.Spec.ClusterRef != nil {
+			return false, errors.Errorf("failed to reconcile cluster reference.  InfraEnv already has cluster reference %s/%s",
+				infraEnv.Spec.ClusterRef.Namespace, infraEnv.Spec.ClusterRef.Name)
+		}
+		var clusterReference aiv1beta1.ClusterReference
+		if err = json.Unmarshal([]byte(clusterReferenceStr), &clusterReference); err != nil {
+			return false, errors.Wrapf(err, "failed to reconcile cluster reference.  Unmarshal of cluster reference %s failed",
+				clusterReferenceStr)
+		}
+		if !reflect.DeepEqual(agent.Spec.ClusterDeploymentName, &clusterReference) {
+			agent.Spec.ClusterDeploymentName = &clusterReference
+			return true, nil
+		}
+	} else if agent.Spec.ClusterDeploymentName != nil {
```

Comment:
Doesn't this effectively prevent users from setting the cluster reference directly on the agent if their agents were discovered using a BMH? This is a pretty large change in behavior given that we already support late-binding with BMH-discovered hosts.

I think we want to keep the existing behavior as well as add this new feature. One option would be to have the annotation present, but with no value, indicate that the agent should not be bound to a cluster. No annotation would then mean that the value on the agent should be respected.

@avishayt what do you think?

Reply:
Changed the removal to depend on annotation on agent which is created when the cluster-reference is initially set.  It means that if the cluster-reference was not set initially by the bmh, it will not be removed if the cluster reference is missing from bmh.

Reply:
If I'm following the question correctly, I think the agent's cluster reference should be set to the value of the BMH annotation when that annotation exists, and not changed when the annotation doesn't exist.

Want to declaratively use the BMH to ensure that the agent is not bound to a cluster? You should make sure the annotation has an empty value; that empty value will be propagated faithfully onto the agent's spec.

We should not be tracking who set what in the past or using that as context for behavioral decisions.